File: DataStructure/Array/498.py
# Diagonal Traverse

from collections import defaultdict


class Solution:
    def findDiagonalOrder(self, mat):
        new_arr = []
        is_upward = True
        out_loop_len = len(mat)
        in_loop_len = len(mat[0])
        i = 0
        j = 0

        while i < out_loop_len and j < in_loop_len:
            new_arr.append(mat[i][j])
            if is_upward:
                if j == in_loop_len - 1:
                    i = i + 1
                    is_upward = not is_upward
                elif i == 0:
                    j = j + 1
                    is_upward = not is_upward
                else:
                    i = i - 1
                    j = j + 1
            else:
                if i == out_loop_len - 1:
                    j = j + 1
                    is_upward = not is_upward
                elif j == 0:
                    i = i + 1
                    is_upward = not is_upward
                else:
                    i = i + 1
                    j = j - 1
        return new_arr


class Solution2:
    def findDiagonalOrder(self, mat):
        d = defaultdict(list)
        out_loop_len = len(mat)
        in_loop_len = len(mat[0])

        for i in range(out_loop_len):
            for j in range(in_loop_len):
                d[i + j].append(mat[i][j])

        ans = []
        for entry in d.items():
            if entry[0] % 2 == 0:
                ans.extend(entry[1][::-1])
            else:
                ans.extend(entry[1])
        return ans

File: DataStructure/Array/498.ts
// Diagonal Traverse

function findDiagonalOrder(mat: number[][]): number[] {
  let newArr: number[] = [];

  const outLoopLen: number = mat.length;
  const inLoopLen: number = mat[0].length;

  let isUpward: boolean = true;
  let i: number = 0;
  let j: number = 0;

  while (i < outLoopLen && j < inLoopLen) {
    newArr.push(mat[i][j]);
    if (isUpward) {
      if (j == inLoopLen - 1) {
        i += 1;
        isUpward = !isUpward;
      } else if (i == 0) {
        j += 1;
        isUpward = !isUpward;
      } else {
        i -= 1;
        j += 1;
      }
    } else {
      if (i == outLoopLen - 1) {
        j += 1;
        isUpward = !isUpward;
      } else if (j == 0) {
        i += 1;
        isUpward = !isUpward;
      } else {
        i += 1;
        j -= 1;
      }
    }
  }
  return newArr;
}

function findDiagonalOrder2(mat: number[][]): number[] {
  const d: { [key: number]: number[] } = {};
  const outLoopLen: number = mat.length;
  const inLoopLen: number = mat[0].length;

  for (let i: number = 0; i < outLoopLen; i++) {
    for (let j: number = 0; j < inLoopLen; j++) {
      const key = i + j;
      if (!d[key]) {
        d[key] = [];
      }
      d[key].push(mat[i][j]);
    }
  }

  const ans: number[] = [];

  Object.keys(d).forEach((key) => {
    const numKey = parseInt(key); // Object.keys의 리턴값이 문자열 타입이기 때문
    if (numKey % 2 === 0) {
      ans.push(...d[numKey].reverse());
    } else {
      ans.push(...d[numKey]);
    }
  });

  return ans;
}
